File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.py
def solution(triangle):
    answer = 0
    l = len(triangle)
    dp = [[] for _ in range(l)]
    dp[0].append(triangle[0][0])
    
    for i in range(1, l):
        for j in range(i+1):
            # 왼쪽
            if j == 0:
                dp[i].append(dp[i-1][0] + triangle[i][j])
            # 오른쪽
            elif j == i:
                dp[i].append(dp[i-1][j-1] + triangle[i][j])
            
            # 가운데
            else:
                dp[i].append(max(dp[i-1][j-1], dp[i-1][j]) + triangle[i][j])
    
    return max(dp[l-1])

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.js
function solution(triangle) {
    var answer = 0;
    let len = triangle.length
    let dp = Array.from({length:len},()=>[])
    for(let i=0;i<len;i++){
        for(let j=0;j<triangle[i].length;j++){
            dp[i].push(-1)
        }
    }
    
    function recur(i, j){
        if(i == len) return 0
        if(j<0 || j>=triangle[i].length) return -999999
        if(dp[i][j]!==-1) return dp[i][j]
        
        dp[i][j]= Math.max(recur(i+1,j), recur(i+1,j+1)) + triangle[i][j]
    
        return dp[i][j]
    }

    recur(0,0)
    return dp[0][0];
}


